File: leetcode-submissions/817_Linked_List_Components.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int numComponents(ListNode* head, vector<int>& nums) {
        unordered_set<int>lookup(nums.begin(), nums.end());
        ListNode* ptr = head;
        int components = 0 ;
        bool isConnected = false;

        while(ptr){
            if(lookup.count(ptr->val)){
                if(not isConnected){
                    components++;
                }
                isConnected= true ;
            }else {
                isConnected = false;
            }
            ptr = ptr->next;
        }
        return components ;
        
    }
};

File: leetcode-submissions/817_Linked_List_Components.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function numComponents(head: ListNode | null, nums: number[]): number {
    // we need to store the values of the nums in the set to lookup in constant time and then we need to check in the linked list values if the consecutive is occurs the store it into the connected component 
    const set:Set<number>=new Set(nums);
    let ptr:ListNode | null =head;
    let components :number =0;
    let isConnnected :boolean =false;
    while(ptr){
        // we need to lookup and check 
    if(set.has(ptr.val)){
        if(!isConnnected){
            components++;
        }
        isConnnected=true;
    }else{
        isConnnected =false;
    }
        ptr=ptr.next;

    }
    return components ;
};